File: CV_2-TRANSFORMATIONS/exercise5.m
% Read images
windmill = imread('windmill.png');
mask = imread('windmill_mask.png');
background = imread('windmill_back.jpeg');

% Convert images to double for processing
windmill = im2double(windmill);
mask = im2double(rgb2gray(mask));
background = im2double(background);

% Invert the mask
mask = ~mask;

% Find bounding box
[row,col] = find(mask);
crop_top = min(row);
crop_bottom = max(row);
crop_left = min(col);
crop_right = max(col);

% Crop images
mask = mask(crop_top:crop_bottom, crop_left:crop_right);
windmill = windmill(crop_top:crop_bottom, crop_left:crop_right, :);

% Resize
new_width = round(size(mask,2)/2);
new_height = round(size(mask,1)/2);
mask = imresize(mask, [new_height new_width]);
windmill = imresize(windmill, [new_height new_width]);
background = imresize(background, 0.5);

% Get dimensions
[bg_height, bg_width, ~] = size(background);
y_offset = round((bg_height - new_height)/2);
x_offset = round((bg_width - new_width)/2);

% Create video writer
v = VideoWriter('transf_windmill_matlab.avi');
v.FrameRate = 30;
open(v);

% Animation parameters
num_frames = 200;
angle_step = -360/num_frames;

% Main loop
for frame = 1:num_frames
    % Calculate rotation angle
    angle = (frame-1) * angle_step;
    
    % Create rotation matrix
    R = [cosd(angle) -sind(angle); sind(angle) cosd(angle)];
    
    % Rotate windmill and mask
    rotated_windmill = imrotate(windmill, angle, 'bilinear', 'crop');
    rotated_mask = imrotate(mask, angle, 'bilinear', 'crop');
    
    % Create frame
    current_frame = background;
    
    % Get ROI
    roi = current_frame(y_offset+1:y_offset+new_height, x_offset+1:x_offset+new_width, :);
    
    % Apply mask
    for c = 1:3
        roi(:,:,c) = rotated_windmill(:,:,c) .* rotated_mask + roi(:,:,c) .* (1-rotated_mask);
    end
    
    % Insert ROI back
    current_frame(y_offset+1:y_offset+new_height, x_offset+1:x_offset+new_width, :) = roi;
    
    % Convert to uint8 before writing
    current_frame = im2uint8(current_frame);
    
    % Write frame
    writeVideo(v, current_frame);
end

close(v);

File: CV_2-TRANSFORMATIONS/exercise5.py
import cv2
import numpy as np

def create_windmill_animation():
    # Read images
    windmill = cv2.imread('windmill.png', cv2.IMREAD_UNCHANGED)
    mask = cv2.imread('windmill_mask.png', cv2.IMREAD_GRAYSCALE)
    background = cv2.imread('windmill_back.jpeg')

    # Invert the mask so the blades are white and background is black
    mask = cv2.bitwise_not(mask)

    # Find the bounding box of the windmill blades to crop
    coords = cv2.findNonZero(mask)
    x, y, w, h = cv2.boundingRect(coords)

    # Crop all images to the bounding box
    mask = mask[y:y + h, x:x + w]
    windmill = windmill[y:y + h, x:x + w]

    # Define smaller dimensions (about 1/2 of the original size to match reference)
    new_width = mask.shape[1] // 2
    new_height = mask.shape[0] // 2

    # Resize all images
    mask = cv2.resize(mask, (new_width, new_height))
    windmill = cv2.resize(windmill, (new_width, new_height))
    background = cv2.resize(background, (background.shape[1] // 2, background.shape[0] // 2))  # Keep background larger

    # Calculate position to place the blades in the center of background
    y_offset = (background.shape[0] - new_height) // 2
    x_offset = (background.shape[1] - new_width) // 2

    # Get dimensions for video
    height, width = background.shape[:2]

    # Create video writer
    fourcc = cv2.VideoWriter_fourcc(*'XVID')
    out = cv2.VideoWriter('transf_windmill.avi', fourcc, 30.0, (width, height))

    # Animation parameters
    num_frames = 200  # Increased number of frames
    angle_step = -360 / num_frames  # Negative for clockwise rotation, smaller step for slower rotation

    # Get center of rotation for the blades
    center = (new_width // 2, new_height // 2)

    for frame in range(num_frames):
        # Calculate rotation angle
        angle = frame * angle_step

        # Create rotation matrix
        M = cv2.getRotationMatrix2D(center, angle, 1.0)

        # Rotate windmill and mask
        rotated_windmill = cv2.warpAffine(windmill, M, (new_width, new_height))
        rotated_mask = cv2.warpAffine(mask, M, (new_width, new_height))

        # Create frame by combining background and rotated windmill
        frame = background.copy()

        # Create ROI in the frame
        roi = frame[y_offset:y_offset + new_height, x_offset:x_offset + new_width]

        # Extract RGB channels from rotated windmill
        windmill_rgb = rotated_windmill[:, :, :3]

        # Create binary mask
        binary_mask = rotated_mask > 0

        # Apply mask to ROI
        for c in range(3):
            roi[:, :, c] = np.where(binary_mask, windmill_rgb[:, :, c], roi[:, :, c])

        # Put ROI back into frame
        frame[y_offset:y_offset + new_height, x_offset:x_offset + new_width] = roi

        # Write frame
        out.write(frame)

    # Release video writer
    out.release()

try:
    create_windmill_animation()
    print("Animation created successfully!")
except Exception as e:
    print(f"An error occurred: {str(e)}")